src/iter/iter.rs

```diff
@@ -58,6 +59,70 @@ impl<'g, K, V> Iterator for Values<'g, K, V> {
     }
 }
 
+/// A draining iterator over a map's entries.
+#[derive(Debug)]
+pub struct Drain<'g, K, V, S>
+where
+    K: Sync + Send + Clone + Eq + Hash,
+    V: Sync + Send,
+    S: BuildHasher,
+{
+    pub(crate) node_iter: NodeIter<'g, K, V>,
+    pub(crate) map: &'g crate::HashMap<K, V, S>,
+    pub(crate) guard: &'g Guard,
+}
+
+impl<'g, K, V, S> Iterator for Drain<'g, K, V, S>
+where
+    K: Sync + Send + Clone + Eq + Hash,
+    V: Sync + Send,
+    S: BuildHasher,
+{
+    type Item = (K, V);
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let node = self.node_iter.next()?;
+
+            if let Some(value) = self.map.remove(&node.key, self.guard) {
+                return Some((node.key.clone(), unsafe {
+                    std::ptr::read(value as *const V)
```

Comment:
You're right, I added this when playing around with giving the user a `V` instead of a `&V`, but we probably should just always give out `&V`s and let them clone when they need to...

Reply:
I'd be nice to have the same signature as `std::HashMap`, but since we're not implementing a trait I guess we can just have the user deal with it...

Reply:
This should be fixed now. I'll write some tests to make sure there actually aren't any double-frees  tomorrow.